### PR TITLE
Split up documentation into subpages and clean up some warnings

### PR DESCRIPTION
**Pull request by mattip**
xref gh-32838, gh-34032

This is a major refactor of parts of the documentation to split it up using sphinx's `autosummary` feature which will build out `autofuction` and `autoclass` stub files and link to them. The end result is that the top module pages like torch.nn.rst and torch.rst are now more like table-of-contents to the actual single-class or single-function documentations pages.

Along the way, I modified many of the docstrings to eliminate sphinx warnings when building. I think the only thing I changed from a non-documentation perspective is to add names to `__all__` when adding them to `globals()` in `torch.__init__.py`

I do not know the CI system: are the documentation build artifacts available after the build, so reviewers can preview before merging?

### CODE REVIEW

File: torch/autograd/functional.py
Comment (albanD) [inline torch/autograd/functional.py:634]: I think you deleted a function here...

Comment (mattip) [reply torch/autograd/functional.py:634]: thanks, reverting

**Comment by ShawnZhong** [comment]
I really like the way you organized the functions to the `generated` folder and the changes you made to the docs. 

FYI, I built the docs based on https://github.com/pytorch/pytorch/pull/37419/commits/2fd0354b1922ecd1612ea17d5ed622d6b77d1d4e, and published it at https://shawnzhong.github.io/pytorch/.

File: torch/onnx/__init__.py
Comment (rgommers) [inline torch/onnx/__init__.py:196]: minor: typo in version here. also `set_training` looks like it needs double backticks

**Review by rgommers — COMMENTED** [review]
This looks quite good and from some testing it seems like this resolves both the searching and the "docs are heavy" issues well. 

Visually I think the vertical margins on every function/class entry in autosummary tables could be made smaller, that way things get more compact and it would read even better. Non-blocking comment though, it's a matter of taste and either way this is a significant improvement.

image

File: torch/onnx/__init__.py
Comment (mattip) [reply torch/onnx/__init__.py:196]: ok, should be 1.6

**Comment by rgommers** [comment]
> I do not know the CI system: are the documentation build artifacts available after the build, so reviewers can preview before merging?

There's a CircleCI job `pytorch_python_doc_push` that builds the docs, however it doesn't store any artifacts for PRs. Controlled by `.circleci/scripts/python_doc_push_script.sh`. I think that's the only doc build (?). Not sure why it doesn't store the artifacts, that would be useful.

File: torch/onnx/__init__.py
Comment (rgommers) [reply torch/onnx/__init__.py:196]: and "version" not "verison"

**Comment by mattip** [comment]
TL DR: there is a broken test. I propose redoing it properly which will require a few days work.

The `test/test_docs_coverage.py` test is failing since it does this:
- scan a top-level module's rst file for `autofunction` directives to determine which functions are documented
- scan `[a.__doc__ for a in dir(module)]` for the string `.. function` in the docstring to determine which functions should be documented
- compare the two, they should be equal.

This PR broke both sides of the equation: 
- `autosummary` replaces the `autofunction` directives and builds out stub files that in turn are linked back in, so now the documentation has a multi-level heirarchy
- the leading `.. function` directive broke `autosummary` so I removed it.

I propose we use [sphinx.ext.coverage](https://www.sphinx-doc.org/en/master/usage/extensions/coverage.html?highlight=coverage) to do this for us, but it means investing time to get the call right and parse the output. Are there other options? Should we add a `make coverage` CI step (which will also require work) to check coverage instead of a test? Should we disable the check entirely for now, and open a new issue to fix it?

**Comment by ShawnZhong** [comment]
I was checking the links for the functions and found that some overloaded functions are not linked correctly in [`torch.html`](https://shawnzhong.github.io/pytorch/torch.html#torch.add)

For those functions, the second column shows the function signature instead of the description, and the link target is set to [torch.html#torch.mul](https://shawnzhong.github.io/pytorch/torch.html#torch.mul) instead of [generated/point/torch.mul.html?highlight=mul](https://shawnzhong.github.io/pytorch/generated/point/torch.mul.html?highlight=mul)

image

File: docs/source/torch.rst
Comment (ShawnZhong) [inline docs/source/torch.rst:238]: Maybe you want to use `generated`?

**Comment by ezyang** [comment]
> I propose we use sphinx.ext.coverage to do this for us, but it means investing time to get the call right and parse the output. Are there other options? Should we add a make coverage CI step (which will also require work) to check coverage instead of a test? Should we disable the check entirely for now, and open a new issue to fix it?

Reading over the sphinx coverage docs, it does seem like this is the right thing to do. You can see the original PR https://github.com/pytorch/pytorch/pull/16039 we mostly added this test script in anger. The important thing is the test needs to fail when you don't document and don't explicitly blacklist.

However, there's a huge benefit to getting this PR in directly, so I think that it is OK not to block this PR and disable the check for now, and then fix it later.

cc @zasdfgbnm who originally added the test

**Comment by zasdfgbnm** [comment]
> > I propose we use sphinx.ext.coverage to do this for us, but it means investing time to get the call right and parse the output. Are there other options? Should we add a make coverage CI step (which will also require work) to check coverage instead of a test? Should we disable the check entirely for now, and open a new issue to fix it?
> 
> Reading over the sphinx coverage docs, it does seem like this is the right thing to do. You can see the original PR https://github.com/pytorch/pytorch/pull/16039 we mostly added this test script in anger. The important thing is the test needs to fail when you don't document and don't explicitly blacklist. The only plausible alternative is to bulk up the script so that it can handle hierarchy on the documentation; probably not too hard either, but if Sphinx has a built in to do this, we probably should use it. 
> 
> However, there's a huge benefit to getting this PR in directly, so I think that it is OK not to block this PR and disable the check for now, and then fix it later.
> 
> cc @zasdfgbnm who originally added the test

I agree that it is a good idea to just disable this test to unblock this. After you disable it, can you open an issue for that?

File: docs/source/torch.rst
Comment (mattip) [reply docs/source/torch.rst:238]: yes, fixed

**Comment by mattip** [comment]
- test disabled. xref issue gh-37590
- fixed (via review comment)
  > some overloaded functions are not linked correctly in torch.html
- rebased to pick up new `vander` function

File: torch/nn/modules/loss.py
Comment (ezyang) [inline torch/nn/modules/loss.py:1294]: Huh, `.. Note::` is not actually a thing?

**Review by ezyang — APPROVED** [review]

File: torch/nn/modules/loss.py
Comment (mattip) [reply torch/nn/modules/loss.py:1294]: `.. note::` is [a directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#paragraph-level-markup) for general RST and rendered as a `<div>` with formatting to look like a box, `Note:` is [a docstring heading](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html?highlight=napoleon#docstring-sections). I think the intent here is a docstring section

File: docs/source/torch.rst
Comment (ShawnZhong) [reply docs/source/torch.rst]: Not sure if it's just me, but it seems that this change doesn't fix the link issue mentioned above: `torch.mul` still links to the [torch.html#torch.mul](https://shawnzhong.github.io/pytorch/torch.html#torch.mul).

File: docs/source/torch.rst
Comment (ShawnZhong) [inline docs/source/torch.rst:194]: I think you might have removed the sub-title? 

Here are the screenshots of the sidebar for torch.html

| Before | After | 
| --- | --- |
| image | image |

File: docs/source/torch.rst
Comment (mattip) [reply docs/source/torch.rst]: Strange. I don't see this on two different builds (both with cuda installed, one on a machine with a GPU, one without). There is nothing obvious in `torch/_torch_docs.py` in the docstrings for these functions as overloaded from `add_docstr(torch.mul,...` and friends. What do you mean by "some overloaded functions" - is there something special about `mul` and `pow` that my builds are not picking up?

How are you building and uploading the docs to https://shawnzhong.github.io/pytorch ?

I guess I could try to replicate the actual `python_doc_push_script.sh` in a docker, does that CI step run with a GPU?

File: docs/source/torch.rst
Comment (mattip) [reply docs/source/torch.rst]: fixed in cab782a

File: docs/source/torch.rst
Comment (ShawnZhong) [reply docs/source/torch.rst]: I use a [GitHub action](https://github.com/ShawnZhong/pytorch/blob/e9e2d6a863f06234c8abd6a1796decefcbb6aafd/.github/workflows/sphinx.yml) to build and push the HTML files to `gh_pages`, which is served by GitHub pages. 

It might be a different environment that causes this issue. I tried to use the same environment as CircileCI, but it seems that I don't have the permission to pull the docker image for building docs. BTW, I just opened a PR to upload the build artifact for python docs at https://github.com/pytorch/pytorch/pull/37658.

Comment (ShawnZhong) [reply docs/source/torch.rst]: Nvm, it works on the CircleCI build. https://5319560-65600975-gh.circle-artifacts.com/0/~/workspace/build_artifacts/master/index.html

**Review by ShawnZhong — APPROVED** [review]

**Comment by ezyang** [comment]
Still working on this.

**Comment by suo** [comment]
Sadly, this breaks pytorch_linux_xenial_cuda10_1_cudnn7_py3_slow_test on master (related to doc breakages). Example: https://app.circleci.com/pipelines/github/pytorch/pytorch/163716/workflows/dbdb10ec-66b9-49b0-9609-6e8968f71c53/jobs/5335902/steps

I'm going to try splitting out doc tests into their own job so that you can get signal earlier on this.

**Comment by suo** [comment]
Reverting :(

**Comment by ezyang** [comment]
@mattip I'm going to take care of relanding this.

**Comment by mattip** [comment]
I am not sure whether to close gh-32838, gh-34032. There is (and probbly will always be) more to do here:
- quantization.html and maybe other pages still could be broken into smaller units.
- remaining warnings should be fixed, and we could turn on sphinx warning-as-error once warnings reach 0 in order to prevent new warnings from appearing
- css and SEO improvements probably better left to web designers:
  - left navbar: subtopic indentation, using standard triangles to collapse topics,
  - right navbar that appears in subtopics should be used in sub-subtopics as well
  - keywords to help search engines
  - robot.txt to prevent search engines finding older versions of the documentation
- I did not look at the site on mobile

**Comment by ezyang** [comment]
I think it's ok to file new issues for these, since the main issue has been addressed.

**Comment by mattip** [comment]
xref
gh-38010 to split more files
gh-38011 to remove warnings
gh-38012 to improve formatting.
I see that searching for "pytorch autograd" already points to the stable documentation, so SEO seems OK.

I think we can close the main issues gh-32838, gh-34032 now.

**Comment by mrshenli** [comment]
Hey @mattip, we noticed that `rpc/index.rst` is removed from this PR. That page is the landing page for the RPC package. Is there any plan for adding it back before v1.6 branch cut?

cc @rohan-varma @jlin27

**Comment by mattip** [comment]
> That page is the landing page for the RPC package

Since much of the content of `docs/source/rpc/index.rst` page was repeated on `docs/source/rpc.rst` I took the liberty of consolidationg them into the latter. If there are external references to the former and not the latter, maybe we should switch it around? It seemed strange to me that there was an `index.rst` page but no other pages. The content is rendered at https://pytorch.org/docs/master/rpc.html.

**Comment by mrshenli** [comment]
Hey @mattip 

Thanks for the info. And the change looks OK to me. Just one concern, it seems like the RPC package disappeared from the main indexing page (and the left navigation bar) on master: https://pytorch.org/docs/master/

Do we plan to fix that?

In the stable release (https://pytorch.org/docs/stable/), there is a `Distributed RPC Framework` link points to the `rpc.rst` page.

**Pull request closed without merging**
